x-pack/plugins/observability/public/context/date_picker_context.tsx

```diff
@@ -0,0 +1,156 @@
+import { useLocation, useHistory } from 'react-router-dom';
+import { parse } from 'query-string';
+import { fromQuery, ObservabilityPublicPluginsStart, toQuery } from '..';
+import { useKibana } from '../../../../../src/plugins/kibana_react/public';
+import { getAbsoluteTime } from '../utils/date';
+
+export interface DatePickerContextValue {
+  relativeStart: string;
+  relativeEnd: string;
+  absoluteStart?: number;
+  absoluteEnd?: number;
+  refreshInterval: number;
+  refreshPaused: boolean;
+  updateTimeRange: (params: { start: string; end: string }) => void;
+  updateRefreshInterval: (params: { interval: number; isPaused: boolean }) => void;
+  lastUpdated: number;
+}
+
+/**
+ * This context contains the time range (both relative and absolute) and the
+ * autorefresh status of the overview page date picker.
+ * It also updates the URL when any of the values change
+ */
+export const DatePickerContext = createContext({} as DatePickerContextValue);
+
+export function DatePickerContextProvider({ children }: { children: React.ReactElement }) {
+  const location = useLocation();
+  const history = useHistory();
+
+  const updateUrl = useCallback(
+    (nextQuery: {
+      rangeFrom?: string;
+      rangeTo?: string;
+      refreshPaused?: boolean;
+      refreshInterval?: number;
+    }) => {
+      history.push({
+        ...location,
+        search: fromQuery({
+          ...toQuery(location.search),
+          ...nextQuery,
+        }),
+      });
+    },
+    [history, location]
+  );
+
+  const [lastUpdated, setLastUpdated] = useState(Date.now());
+
+  const { data } = useKibana<ObservabilityPublicPluginsStart>().services;
+
+  const defaultTimeRange = data.query.timefilter.timefilter.getTimeDefaults();
+  const sharedTimeRange = data.query.timefilter.timefilter.getTime();
+  const defaultRefreshInterval = data.query.timefilter.timefilter.getRefreshIntervalDefaults();
+  const sharedRefreshInterval = data.query.timefilter.timefilter.getRefreshInterval();
+
+  const {
+    rangeFrom = sharedTimeRange.from ?? defaultTimeRange.from,
+    rangeTo = sharedTimeRange.to ?? defaultTimeRange.to,
+    refreshInterval = sharedRefreshInterval.value || defaultRefreshInterval.value || 10000, // we want to override a default of 0
+    refreshPaused = sharedRefreshInterval.pause ?? defaultRefreshInterval.pause,
+  } = parse(location.search, {
+    sort: false,
+  });
+
+  const relativeStart = rangeFrom as string;
+  const relativeEnd = rangeTo as string;
+
+  const absoluteStart = useMemo(
+    () => getAbsoluteTime(relativeStart),
+    // `lastUpdated` works as a cache buster
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [relativeStart, lastUpdated]
+  );
+
+  const absoluteEnd = useMemo(
+    () => getAbsoluteTime(relativeEnd, { roundUp: true }),
+    // `lastUpdated` works as a cache buster
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [relativeEnd, lastUpdated]
+  );
```

Comment:
If `getAbsoluteTime` is a pure function, and looks like it is, providing `lastUpdated` as a dependency here serves no purpose and might be confusing to a reader.

`lastUpdated` provided below on L#125 is already making the context to refresh.

Reply:
tl;dr The purpose of `lastUpdate` is to bust the cached `absolute*` values only when the user interacts with the refresh button, even if the relative values don't change.

----

Consider these values:
```
relativeStart = 'now-15m';
relativeEnd = 'now';
```

We need to resolve these to absolute values (`Date.now() - 15_minutes` and `Date.now()`). If we remove the `lastUpdate` the code ends up like this:

```ts
// Option 1
const absoluteStart = useMemo(
  () => getAbsoluteTime(relativeStart),
  [relativeStart]
);

const absoluteEnd = useMemo(
  () => getAbsoluteTime(relativeEnd, { roundUp: true }),
  [relativeEnd]
);
```

This looks fair, right? But, it doesn't work. It doesn't update the values if the user clicks on the "Refresh" button (the bug this PR is fixing). When the user clicks the button, the `relative*` values stay the same: `now-15m` and `now`. The `absolute*` values then remain frozen unless the user changes both.

We could remove the `useMemo` altogether and then the `absolute*` values will always update.

```ts
// Option 2
const absoluteStart = getAbsoluteTime(relativeStart);
const absoluteEnd = getAbsoluteTime(relativeEnd, { roundUp: true }));
```

This is [what the previous code did](https://github.com/elastic/kibana/blob/61a84edc622e376120b0049b575cfaa394c95e35/x-pack/plugins/observability/public/hooks/use_time_range.ts#L38-L39). The issue here is that the absolute values resolve _in every render_. This is not great. A change in the absolute values will trigger a refetch, which will trigger a re-render, which will trigger a change in the absolute values, which will trigger a refetch, which will trigger a re-render, which will trigger a change in the absolute values, which will... you get the idea.

The way this was resolved was [to use the `relative*` values](https://github.com/elastic/kibana/blob/61a84edc622e376120b0049b575cfaa394c95e35/x-pack/plugins/observability/public/components/app/section/logs/index.tsx#L72-L74) to trigger a refetch, but as I have explained earlier, the relative values don't change when the user clicks "Refresh", so the button wasn't working.

What I do is cache the value with `useMemo` to prevent the `absolute*` values from updating all the time, but use a flag (`lastUpdated`) that ensures the resolved values update [when the user interacts with the button](https://github.com/elastic/kibana/pull/126927/files#diff-d23c72fd09d766504bc30153dbb87c5b690619ca1007534bee284b7790d98af6R96). 

----

How could I make this clearer in code? Did you the name `lastUpdated` not clear enough? Should I bring more attention to the comment over the hook dependencies?

Reply:
My apologies @afgomez, I didn't realize that `datemath.parse` called by `getAbsoluteTime` makes `getAbsoluteTime`  an impure function.

Reply:
Do we need the `absolute*` values to be cached? `lastUpdated` is also a dependency in the `useFetcher` hook so it would trigger a refetch, so we could resolve the `absolute*` values then. I think this will make clear when we have to fetch the data.
What do you think?

Reply:
The cache is there to keep the `absolute*` values stable. If we don't memoize them they will change on each render.

We could prevent that resolving the values on `useFetcher`, but then each section will resolve their own `absolute*` values, since they run at different times (only within milliseconds within each other, which may or may not matter).